File: src/main/scala/gov/ornl/stucco/rt/topology/Topology.scala
package gov.ornl.stucco.rt.topology

import gov.ornl.stucco.rt.spout._
import gov.ornl.stucco.rt.bolt._

import org.streum.configrity._
import org.streum.configrity.yaml._

import grizzled.slf4j.Logging

import backtype.storm.{Config, LocalCluster, StormSubmitter}
import backtype.storm.topology.TopologyBuilder
import backtype.storm.utils.Utils
import backtype.storm.spout.Scheme

import com.rapportive.storm.spout.AMQPSpout

/** An object to define and run the storm topology.
  *
  * This object defines the topology, and then it runs it either
  * locally or remotely.
  *
  * This object builds the storm topology using a `TopologyBuilder`.
  * Given an IP address for the Nimbus server, it submits the topology
  * using `StormSubmitter`. Otherwise, it runs the topology locally using
  * a `LocalCluster`.
  */
object Topology extends Logging {

  /** Location of the settings file.
    */
  private val settingsFile = "config.yaml"

  private val settings = Configuration.load("config.yaml", YAMLFormat)

  /** Create and run the storm topology.
    *
    * @param args Empty array or an array containing a single string - the
    * IP address of the Nimbus machine.
    */
  def main(args: Array[String]) {
    val builder = new TopologyBuilder
    buildTopology(builder)

    val config = new Config
    config.setDebug(settings[Boolean]("storm.debug"))

    if (args.nonEmpty) {
      config.setNumWorkers(settings[Int]("storm.workers"))
      StormSubmitter.submitTopology(args(0), config, builder.createTopology())
    } else {
      val cluster = new LocalCluster
      cluster.submitTopology("Topology", config, builder.createTopology())
      // a storm topology runs forever (until it's killed)
      while (true) {
        Utils.sleep(5000)
      }
      cluster.killTopology("Topology")
      cluster.shutdown()
    }
  }

  /** Build the topology.
    *
    * Builds the entire topology for the computation.
    *
    * @param builder The `TopologyBuilder` to use.
    *
    * @see `topology.txt` - a ASCII art graphical depiction of the topology
    */
  def buildTopology(builder: TopologyBuilder) {
    val spout = buildSpout(new Deserializer)
    // build spout
    builder.setSpout("rabbitmq", spout, settings[Int]("instances.rabbitmq"))
    builder.setBolt("uuid", new UUIDBolt, settings[Int]("instances.uuid"))
      .shuffleGrouping("rabbitmq")
    builder.setBolt("route", new RouteBolt, settings[Int]("instances.route"))
      .shuffleGrouping("uuid")
    // build structured bolts
    builder.setBolt("parse", new ParseBolt, settings[Int]("instances.parse"))
      .shuffleGrouping("route", "structured")
    builder.setBolt("split", new SplitBolt, settings[Int]("instances.split"))
      .shuffleGrouping("parse")
    builder.setBolt("structuredgraph", new StructuredGraphBolt, settings[Int]("instances.structuredgraph"))
      .shuffleGrouping("split")
    // build unstructured bolts
    builder.setBolt("extract", new ExtractBolt, settings[Int]("instances.extract"))
      .shuffleGrouping("route", "unstructured")
    builder.setBolt("concept", new ConceptBolt, settings[Int]("instances.concept"))
      .shuffleGrouping("extract")
    builder.setBolt("relation", new RelationBolt, settings[Int]("instances.relation"))
      .shuffleGrouping("concept")
    builder.setBolt("unstructuredgraph", new UnstructuredGraphBolt, settings[Int]("instances.unstructuredgraph"))
      .shuffleGrouping("relation")
    // build common bolts
    builder.setBolt("document", new DocumentBolt, settings[Int]("instances.document"))
      .shuffleGrouping("parse").shuffleGrouping("extract")
    builder.setBolt("graph", new GraphBolt, settings[Int]("instances.graph"))
      .shuffleGrouping("structuredgraph").shuffleGrouping("unstructuredgraph")
  }

  /** Build the spout.
    *
    * Constructs a RabbitMQ queue and using a given data deserialization scheme,
    * creates a RabbitMQ spout.
    *
    * @param scheme The scheme to use to deserialize binary data from the queue.
    */
  def buildSpout(scheme: Scheme) = {
    val name = settings[String]("rabbitmq.queue.name")
    val durable = settings[Boolean]("rabbitmq.queue.durable")
    val exclusive = settings[Boolean]("rabbitmq.queue.exclusive")
    val autoDelete = settings[Boolean]("rabbitmq.queue.autoDelete")

    val queue = SimpleQueue(name, durable, exclusive, autoDelete)

    val host = settings[String]("rabbitmq.host")
    val port = settings[Int]("rabbitmq.port")
    val username = settings[String]("rabbitmq.username")
    val password = settings[String]("rabbitmq.password")
    val vhost = settings[String]("rabbitmq.vhost")

    new AMQPSpout(host, port, username, password, vhost, queue, scheme)
  }
}

File: stucco-topology/src/main/java/gov/ornl/stucco/topology/Topology.java
package gov.ornl.stucco.topology;

import gov.ornl.stucco.bolt.AlignmentBolt;
import gov.ornl.stucco.bolt.ConceptBolt;
import gov.ornl.stucco.bolt.ExtractBolt;
import gov.ornl.stucco.bolt.ParseBolt;
import gov.ornl.stucco.bolt.RelationBolt;
import gov.ornl.stucco.bolt.UUIDBolt;
import gov.ornl.stucco.spout.RabbitMQTopicSpout;

import java.util.List;
import java.util.Map;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.yaml.snakeyaml.Yaml;

import backtype.storm.Config;
import backtype.storm.LocalCluster;
import backtype.storm.StormSubmitter;
import backtype.storm.generated.AlreadyAliveException;
import backtype.storm.generated.InvalidTopologyException;
import backtype.storm.topology.TopologyBuilder;


public class Topology {
	private static final Logger logger = LoggerFactory.getLogger(Topology.class);
	private static final String CONFIG_FILE_NAME = "config.yaml";
	private static final int DEFAULT_INSTANCES = 2;
	
	private enum BOLTS {
		STRUCTURED_DATA,
		UNSTRUCTURED_DATA,
		UUID_STRUCT,
		PARSE,
		UUID_UNSTRUCT,
		EXTRACT,
		CONCEPT,
		RELATION,
		ALIGNMENT
	}
	
	public static final String EMPTY_GRAPHSON = "{ \"edges\":[], \"vertices\":[] }";
	public static final String DOC_SERVICE_CLIENT_HOST = "doc_client_host";
	public static final String DOC_SERVICE_CLIENT_PORT = "doc_client_port";
	public static final int LOCAL_MODE_PARALLELISM = 2;
	
	private TopologyBuilder builder;
	private Map<String, Object> configMap;
	
	@SuppressWarnings("unchecked")
	public Topology() {
		builder = new TopologyBuilder();
		
		Yaml yaml = new Yaml();
		//read in yaml file as json object
		configMap = (Map<String, Object>) yaml.load(Topology.class.getClassLoader().getResourceAsStream(CONFIG_FILE_NAME));
		Integer structSpoutInstances = DEFAULT_INSTANCES;
		Integer structUuidInstances = DEFAULT_INSTANCES;
		Integer parseInstances = DEFAULT_INSTANCES;
		Integer unstructSpoutInstances = DEFAULT_INSTANCES;
		Integer unstructUuidInstances = DEFAULT_INSTANCES;
		Integer extractInstances = DEFAULT_INSTANCES;
		Integer conceptInstances = DEFAULT_INSTANCES;
		Integer realtionInstances = DEFAULT_INSTANCES;
		Integer alignmentInstances = DEFAULT_INSTANCES;
		try {
			structSpoutInstances = ((Map<String, Integer>) configMap.get("instances")).get(BOLTS.STRUCTURED_DATA.toString().toLowerCase());
			structUuidInstances = ((Map<String, Integer>) configMap.get("instances")).get(BOLTS.UUID_STRUCT.toString().toLowerCase());
			parseInstances = ((Map<String, Integer>) configMap.get("instances")).get(BOLTS.PARSE.toString().toLowerCase());
			unstructSpoutInstances = ((Map<String, Integer>) configMap.get("instances")).get(BOLTS.UNSTRUCTURED_DATA.toString().toLowerCase());
			unstructUuidInstances = ((Map<String, Integer>) configMap.get("instances")).get(BOLTS.UUID_UNSTRUCT.toString().toLowerCase());
			extractInstances = ((Map<String, Integer>) configMap.get("instances")).get(BOLTS.EXTRACT.toString().toLowerCase());
			conceptInstances = ((Map<String, Integer>) configMap.get("instances")).get(BOLTS.CONCEPT.toString().toLowerCase());
			realtionInstances = ((Map<String, Integer>) configMap.get("instances")).get(BOLTS.RELATION.toString().toLowerCase());
			alignmentInstances = ((Map<String, Integer>) configMap.get("instances")).get(BOLTS.ALIGNMENT.toString().toLowerCase());
		} catch (Exception ex) {
			logger.warn("Could not find instances in config file, defaulting to '" + DEFAULT_INSTANCES + "' instances.");
		}
		
		//build the topology
		//create structured data topic spout
		//TODO check for non-null spout
		RabbitMQTopicSpout structSpout = buildSpout(BOLTS.STRUCTURED_DATA.toString().toLowerCase());
		builder.setSpout(BOLTS.STRUCTURED_DATA.toString().toLowerCase(), structSpout, structSpoutInstances);
		logger.debug("spout [" + BOLTS.STRUCTURED_DATA.toString().toLowerCase() + "] built with " + structSpoutInstances + " instances.");
		
		builder.setBolt(BOLTS.UUID_STRUCT.toString().toLowerCase(), (new UUIDBolt()), structUuidInstances).shuffleGrouping(BOLTS.STRUCTURED_DATA.toString().toLowerCase());
		logger.debug("bolt [" + BOLTS.UUID_STRUCT.toString().toLowerCase() + "] built with " + structUuidInstances + " instances.");
		
		builder.setBolt(BOLTS.PARSE.toString().toLowerCase(), (new ParseBolt()), parseInstances).shuffleGrouping(BOLTS.UUID_STRUCT.toString().toLowerCase());
		logger.debug("bolt [" + BOLTS.PARSE.toString().toLowerCase() + "] built with " + parseInstances + " instances.");
		
		//create unstructured data topic spout
		//TODO check for non-null spout
		RabbitMQTopicSpout unstructSpout = buildSpout(BOLTS.UNSTRUCTURED_DATA.toString().toLowerCase());
		builder.setSpout(BOLTS.UNSTRUCTURED_DATA.toString().toLowerCase(), unstructSpout, unstructSpoutInstances);
		logger.debug("spout [" + BOLTS.UNSTRUCTURED_DATA.toString().toLowerCase() + "] built with " + unstructSpoutInstances + " instances.");
		
		builder.setBolt(BOLTS.UUID_UNSTRUCT.toString().toLowerCase(), (new UUIDBolt()), unstructUuidInstances).shuffleGrouping(BOLTS.UNSTRUCTURED_DATA.toString().toLowerCase());
		logger.debug("bolt [" + BOLTS.UUID_UNSTRUCT.toString().toLowerCase() + "] built with " + unstructUuidInstances + " instances.");
		
		builder.setBolt(BOLTS.EXTRACT.toString().toLowerCase(), (new ExtractBolt()), extractInstances).shuffleGrouping(BOLTS.UUID_UNSTRUCT.toString().toLowerCase());
		logger.debug("bolt [" + BOLTS.EXTRACT.toString().toLowerCase() + "] built with " + extractInstances + " instances.");
		
		builder.setBolt(BOLTS.CONCEPT.toString().toLowerCase(), (new ConceptBolt()), conceptInstances).shuffleGrouping(BOLTS.EXTRACT.toString().toLowerCase());
		logger.debug("bolt [" + BOLTS.CONCEPT.toString().toLowerCase() + "] built with " + conceptInstances + " instances.");
		
		builder.setBolt(BOLTS.RELATION.toString().toLowerCase(), (new RelationBolt()), realtionInstances).shuffleGrouping(BOLTS.CONCEPT.toString().toLowerCase());
		logger.debug("bolt [" + BOLTS.RELATION.toString().toLowerCase() + "] built with " + realtionInstances + " instances.");
		
		builder.setBolt(BOLTS.ALIGNMENT.toString().toLowerCase(), (new AlignmentBolt()), alignmentInstances).shuffleGrouping(BOLTS.PARSE.toString().toLowerCase()).shuffleGrouping(BOLTS.RELATION.toString().toLowerCase());
		logger.debug("bolt [" + BOLTS.ALIGNMENT.toString().toLowerCase() + "] built with " + alignmentInstances + " instances.");
	}
	
	
	@SuppressWarnings("unchecked")
	public RabbitMQTopicSpout buildSpout(String spoutName) {
		RabbitMQTopicSpout spout = null;
		if (configMap.containsKey(spoutName)) {
			try {
				String exchangeName = ((Map<String, String>) configMap.get(spoutName)).get("exchange");
				String host = ((Map<String, String>) configMap.get(spoutName)).get("host");
				Integer port = ((Map<String, Integer>) configMap.get(spoutName)).get("port");
				String username = ((Map<String, String>) configMap.get(spoutName)).get("username");
				String password = ((Map<String, String>) configMap.get(spoutName)).get("password");
				String queueName = ((Map<String, String>) configMap.get(spoutName)).get("queue");
				List<String> bindings = ((Map<String, List<String>>) configMap.get(spoutName)).get("bindings");
				String[] bindingKeys = new String[bindings.size()];
				bindingKeys = bindings.toArray(bindingKeys);
				spout = new RabbitMQTopicSpout(exchangeName, queueName, host, port, username, password, bindingKeys);
                
				logger.debug("SPOUT CREATED on " + exchangeName + " - " + queueName + " with bindings '" + bindingKeys + "'");
			} catch (Exception ex) {
				logger.error("Error reading config file.", ex);
			}
		}
		return spout;
	}
	
	
	@SuppressWarnings("unchecked")
	public void runTopology(String name) {
		//set up the storm configuration
		Boolean debug = ((Map<String, Boolean>) configMap.get("storm")).get("debug");
		if (debug == null) {
			debug = Boolean.TRUE;
		}
		Config config = new Config();
		config.setDebug(debug);
		config.setMessageTimeoutSecs(21600); //TODO: read this from config
		
		if (configMap.containsKey("document_service")) {
			String host = ((Map<String, String>) configMap.get("document_service")).get("host");
			Integer port = ((Map<String, Integer>) configMap.get("document_service")).get("port");
			config.put(DOC_SERVICE_CLIENT_HOST, host);
			config.put(DOC_SERVICE_CLIENT_PORT, port.intValue());
		}
		
		logger.debug("DEBUG MODE is on.");

		if (name != null) {
			//don't bother with worker number and childOpts if using localCluster
			Integer workers = ((Map<String, Integer>) configMap.get("storm")).get("workers");
			if (workers == null) {
				workers = new Integer(DEFAULT_INSTANCES);
			}
			logger.debug("Preparing to submit topology with configuration of " + workers + " worker instances.");
			config.setNumWorkers(workers.intValue());

			String childOpts = ((Map<String, String>) configMap.get("storm")).get("childOpts");
			if (childOpts != null) {
				logger.debug("Settiing childOpts of: " + childOpts);	
				config.put(Config.WORKER_CHILDOPTS, childOpts);
			}else{
				logger.debug("no childOpts specified in config.");	
			}

			try {
				StormSubmitter.submitTopology(name, config, builder.createTopology());
			} catch (AlreadyAliveException e) {
				logger.error("Topology is already running.", e);
			} catch (InvalidTopologyException e) {
				logger.error("Attempted to submit an invalid topology.", e);
			}
		}
		else {
			config.put(Config.TOPOLOGY_MAX_TASK_PARALLELISM, LOCAL_MODE_PARALLELISM);
			LocalCluster localCluster = new LocalCluster();
			localCluster.submitTopology("RT-Topology", config, builder.createTopology());
		}
	}


	/**
	 * @param args
	 */
	public static void main(String[] args) {
		Topology topo = new Topology();
		//TODO: handle args, only pass name here
		topo.runTopology(null);
	}

}
